Add button to download puzzle file

BredBeanQueen opened the pull request:
Add a button in the top right corner of a puzzle view allowing for user to download the .puzzle file directly instead of going through steam.

Currently needs line 116 of web/src/views/om/puzzles/PuzzleView.tsx to be updated to link to a download of the puzzle, instead of the steam page.

BredBeanQueen committed:
Create new button

Duplicate for Get on steam, and list it as "Get Puzzle Direct"
The link target for this is incorrect and would need to be updated

BredBeanQueen committed:
Rename additional button

Now says "Download Puzzle" instead of "Get Puzzle Direct"

Also name the tag that this button and the get on steam button reside in to "Download_Buttons" instead of being a nameless tag
Note this tag is necessary to contain multiple button elements here

12345ieee commented:
I will get to this, but not this weekend.

12345ieee commented:
Please don't keep the branch up to date manually by recursive merging.

Either rebase it cleanly every once in a while or leave it as is.

I've force pushed to your branch such a rebase while I test it.

12345ieee commented:
Compile fails with:
> [eslint] 
src/views/om/puzzles/PuzzleView.tsx
  Line 105:22:  'Download_Buttons' is not defined  react/jsx-no-undef

I'm 100% lost in this.

BredBeanQueen committed:
Correct fragment syntax

When changing the name of the button I also put names into the fragment that allows having multiple element here. However this is then incorrect syntax as a fragment must be an empty element tag or import react and use <Fragment></Fragment>.

This change reverts the Fragment to <></>.

BredBeanQueen commented:
> Please don't keep the branch up to date manually by recursive merging.
> 
> Either rebase it cleanly every once in a while or leave it as is.
> 
> I've force pushed to your branch such a rebase while I test it.

Sorry about this, I was unaware of the correct way to handle this using github, will rebase properly in the future if needed.



> Compile fails with:
> 
> > [eslint]
> > src/views/om/puzzles/PuzzleView.tsx
> > Line 105:22:  'Download_Buttons' is not defined  react/jsx-no-undef
> 
> I'm 100% lost in this.

That is my own fault, didn't realise the syntax of a fragment required the element tag to be specifically <></> when changing it and didn't check that change with an ide. The point of the fragment is to allow having 2 buttons there since that is 2 elements when typescript only allows 1.

12345ieee commented:
I've fixed the display, but I wasn't able to fix the link to hit the api properly, here's the block I made:
```js
                    <Box sx={{ display: "flex", gap: "0.5rem", marginLeft: "auto" }}>
                        <Button
                            size="small"
                            variant="outlined"
                            color="primary"
                            style={{
                                marginTop: "auto",
                                marginBottom: "auto",
                            }}
                        >
                            <a
                                href={`/om/puzzle/${puzzleId}/file`}
                                style={{
                                    color: "inherit",
                                    textDecoration: "none",
                                }}
                            >
                                Download Puzzle
                            </a>
                        </Button>
                        <Button
                            size="small"
                            variant="outlined"
                            color="primary"
                            style={{
                                marginTop: "auto",
                                marginBottom: "auto",
                            }}
                        >
                            <a
                                href={`https://steamcommunity.com/sharedfiles/filedetails/?id=${puzzleId.substring(1)}`}
                                target="_blank"
                                rel="noreferrer"
                                style={{
                                    color: "inherit",
                                    textDecoration: "none",
                                }}
                            >
                                Get on Steam
                            </a>
                        </Button>
                    </Box>
```
